### ISSUE TITLE
Bad S3 Backend Configuration Cannot Be Fixed 

### ISSUE THREAD

**Issue by SJrX** [hashicorp/terraform#23532]
### Terraform Version

```
Terraform v0.12.16
```

### Terraform Configuration Files

```hcl
terraform {
  backend "s3" {
    bucket = "good-bucket-name"
   }
}
```

### Debug Output

```
<Error><Code>NoSuchBucket</Code><Message>The specified bucket does not exist</Message><BucketName>bad-bucket-name</BucketName><RequestId>XXXXXX</RequestId></Error>
2019/11/30 11:30:59 [DEBUG] [aws-sdk-go] DEBUG: Validate Response s3/ListObjects failed, attempt 0/5, error NoSuchBucket: The specified bucket does not exist
        status code: 404, request id: XXXXXX, 
```
### Expected Behavior

Successful initialization

### Actual Behavior
```
Initializing the backend...

Error: Failed to get existing workspaces: S3 bucket does not exist.

The referenced S3 bucket must have been previously created. If the S3 bucket
was created within the last minute, please wait for a minute or two and try
again.

Error: NoSuchBucket: The specified bucket does not exist
```

### Steps to Reproduce

1. Create a terraform backend configuration using `bad-bucket-name`
2. Run `terraform init` which gets an error about the bucket not existing because it's wrong (say a typo).
3. Change terraform backend configuration using `good-bucket-name`
4. Run `terraform init` 

### Additional Context
I'm just learning terraform and mistyped the bucket information, unfortunately the error makes it seem like this could happen naturally, internally it looks like the project is half initialized with the with the bad bucket name (I can delete the .terraform folder to fix it) , even though it never successfully completed. 

### References
N/A

**Comment by pkolyvas** [comment]
@SJrX Thanks for reporting. I was able to reproduce it.

**Comment by jbardin** [comment]
Hi @SJrX,

The way to override the stored backend configuration is to use the `-reconfigure` flag with `init`.
Optionally removing the entire `.terraform` directory would work as well, but require downloading all providers and modules again.

**Comment by SJrX** [comment]
Hey @jbardin , I was slightly confused with your message. The problem isn't that I'm stuck, it's that when you initialize terraform with a bad s3 bucket as a backend, you can't fix it, and the error message is not obvious that it's terraforms state that is the problem. 

Imho terraform should not consider itself initialized if the first initialization failed, or maybe more specifically (again I'm new to terraform), it shouldn't consider the current s3 backend as the source of data that needs to be migrated if it never sucessfully used it.

**Comment by jbardin** [comment]
Hi @SJrX,

Unfortunately in this case, terraform cannot differentiate between a typo in the backend configuration and incorrect credentials or authorization. Since remote state is often shared, there's no way to determine what constitutes the first initialization, as most use cases already have it initialized. 

The fix in this case is to use the `-reconfigure` flag documented in the [Backend Initialization](https://www.terraform.io/docs/commands/init.html#backend-initialization) section, which was added for this express purpose.

We do have some future work pending to restructure how backends operate in general, which will give some room to refactor this UI and provide better feedback for users.

**Comment by griffinator76** [comment]
We ran into this same issue and spent quite a while trying to work out what was happening. 

One issue we found is that Terraform does not log anything to it's log file when this happens (i.e. if TF_LOG and TF_LOG_PATH environment variables are set). The error is output to the console but not recorded in the log, and crucially it does not log the name of the bucket it is trying to use - which would obviously be very useful in identifying that an invalid name was being used. Nothing is logged for this error even when TF_LOG=TRACE.

Based on the fix mentioned above, I think it would also be helpful if this fix was suggested in the error message i.e. run `terraform init -reconfigure` to try and fix this.
